pennylane/tape/tape.py

```diff
@@ -1557,7 +1557,21 @@ def copy(self, copy_operations=False):
             tape._ops = self._ops.copy()
             tape._measurements = self._measurements.copy()
 
-        tape._update()
+        tape._graph = None
+        tape._specs = None
+        tape._depth = None
+
+        tape.wires = copy.copy(self.wires)
+        tape.num_wires = self.num_wires
+        tape.is_sampled = self.is_sampled
+        tape.all_sampled = self.all_sampled
+
+        tape._update_par_info()
+
+        tape._update_trainable_params()
+        tape._obs_sharing_wires = [copy.copy(shared_wires) for shared_wires in self._obs_sharing_wires]
+        tape._obs_sharing_wires_id = [copy.copy(wire_id) for wire_id in self._obs_sharing_wires_id]
+
```

Comment:
This introduces some code duplication but avoids the rather costly parts of _update() that are just recomputing things that can be copied from self. Someone needs to check whether I am doing the copying here on the right level of depth.

Reply:
A good catch! Copying looks correct too. I may suggest to keep the changes within the `_update()` method instead of expanding it here. Do you have a performance report for these changes?

Reply:
Would be good to discuss the application in which the `tape.copy` time is considerate!  For a tape with about 50 operations on 5 qubits (although the latter should not matter too much) the time spent on copying tapes seems to lie below a percent or so.

Reply:
It was considerable while there still was this "bug" that caused O(num_params*num_observables) many take copy/execute operations. It is probably not so bad anymore. Still, when you "copy", I think, you copy everything. I remember that the previous implementation did not copy everything and there were some relatively expensive intermediates (I think _obs_sharing_wires and _obs_sharing_wires_id) that then had to be re-computed in the incomplete copy. Since they are known it seems better to copy them over from the original tape.

Reply:
This seems to be implemented [now in the `QuantumScript` class](https://github.com/PennyLaneAI/pennylane/blob/d81e1a8a2bc26ed83c304ac641037f44592c2723/pennylane/tape/qscript.py#L991) (and before that in the `QuantumTape` class)